core/src/main/java/io/crate/types/IntegerType.java

```diff
@@ -70,12 +70,17 @@ public Integer value(Object value) {
         if (value instanceof BytesRef) {
             return Integer.parseInt(((BytesRef) value).utf8ToString());
         }
-
         long longVal = ((Number) value).longValue();
         if (longVal < Integer.MIN_VALUE || Integer.MAX_VALUE < longVal) {
             throw new IllegalArgumentException("integer value out of range: " + longVal);
         }
-        return ((Number) value).intValue();
+        Integer intValue = ((Number) value).intValue();
+        if (value instanceof Double && intValue.doubleValue() != (double) value) {
+            throw new IllegalArgumentException("Loss of precision for this double");
```

Comment:
I think I'd prefer to have our custom exception for that, so that here: https://github.com/crate/crate/pull/7466/files#diff-64b4c3e3d216c4b9fe749727e9938ba3R173 is more obvious that we had loss of precision when trying to cast.

Reply:
Would `ConversionException` work?

Reply:
Actually, not an option because it is in `sql` and the types are in `core`. Moving it to `core` is tricky because it depends on the `SqlPrinter`.